master/buildbot/steps/artifact.py

```diff
@@ -402,6 +482,133 @@ def start(self):
         self.setCommand(command)
         ShellCommand.start(self)
 
+    @defer.inlineCallbacks
+    def _getBuildRequest(self):
+        triggeredbybrid = self.build.requests[0].id
+        br = yield self.master.db.buildrequests.getBuildRequestTriggered(triggeredbybrid, self.artifactBuilderName)
+        defer.returnValue(br)
+
+
+class DownloadArtifactsFromChildren(LoggingBuildStep, CompositeStepMixin):
+    name = "Download Artifact(s) from triggered builds"
+    description="Downloading artifact(s) from triggered builds..."
+    descriptionDone="Downloaded artifact(s) from triggered builds"
+
+    def __init__(self,
+                 artifactServer,
+                 artifactServerDir,
+                 artifactBuilderName,
+                 workdir='',
+                 artifactServerPort=None,
+                 artifactDestination=None,
+                 artifactDirectory=None,
+                 artifact=None,
+                 usePowerShell=True,
+                 **kwargs):
+        self.workdir = workdir
+        self.artifact = artifact
+        self.artifactBuilderName = artifactBuilderName
+        self.artifactDirectory = artifactDirectory
+        self.artifactServer = artifactServer
+        self.artifactServerDir = artifactServerDir
+        self.artifactServerPort = artifactServerPort
+        self.artifactDestination = artifactDestination
+        self.master = None
+        self.usePowerShell = usePowerShell
+        LoggingBuildStep.__init__(self, **kwargs)
+
+    @defer.inlineCallbacks
+    def start(self):
+        if self.master is None:
+            self.master = self.build.builder.botmaster.parent
+
+        self._setUpLogs()
+
+        partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(self.build.requests[0].id, self.artifactBuilderName)
+        buildRequetsIdsWithArtifacts = self._getBuildRequestIdsWithArtifacts(partitionRequests)
+        self.partitionCount = len(buildRequetsIdsWithArtifacts)
+        self.step_status.setText(["Downloading artifacts from %d triggered partitions" % self.partitionCount])
+        artifactsMap = {}
+        self.build.setProperty("artifactsMap", {})
+        for brid in buildRequetsIdsWithArtifacts:
+            buildRequest = yield self.master.db.buildrequests.getBuildRequestById(brid)
+
+            localdir = self._getLocalDir(brid)
```

Comment:
Yes it adds the build request id to the path, in order to separate the downloaded artifacts. For example if we download artifacts from build; `1`, `3`, `7`, we want the artifacts from those 3 build to be in separate locations, eg. subfolder `1`, `3` and `7` so they dont overwrite each others artifacts.